<a href="https://colab.research.google.com/github/sukhan220/nlp_bangla_notebook/blob/main/tokenizer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# টোকেনাইজেশন ও এনএলপির মৌলিক ধারণা

---

ন্যাচারাল ল্যাঙ্গুয়েজ প্রসেসিং (NLP)-এর প্রথম ও অন্যতম গুরুত্বপূর্ণ ধাপ হলো টেক্সট 'নরমালাইজেশন'। কম্পিউটার যেহেতু সংখ্যা ছাড়া অন্য কোনো তথ্য সরাসরি প্রসেস করতে পারে না, তাই মানুষের ভাষাকে কম্পিউটারের বোধগম্য করতে বাক্যের শব্দগুলোকে সঠিকভাবে এনকোড করতে হয়। কোনো টেক্সট বা বাক্যকে অর্থপূর্ণ ছোট ছোট টুকরো বা শব্দে ভাগ করে সেগুলোকে সংখ্যায় রূপান্তর (Encode) করার এই প্রক্রিয়াকেই বলা হয় টোকেনাইজেশন।

### এনএলপির প্রধান দুটি মূল নীতি (Core Principles)

1. **টোকেনাইজেশন (Tokenization):** টেক্সটকে ভেঙে ছোট ছোট টুকরো বা শব্দে (Token) রূপান্তর করা।
2. **সিকোয়েন্সিং (Sequencing):** সেই শব্দগুলোকে একটি নির্দিষ্ট ধারাবাহিক সংখ্যায় সাজিয়ে কম্পিউটারের জন্য প্রসেস করতে পারে মতো করা।

---

### ক্যারেক্টার বনাম টোকেন এনকোডিং

ভাষার সবচেয়ে ছোট একক হলো 'অক্ষর' বা ক্যারেক্টার (যেমন: ইংরেজিতে ASCII বা বাংলায় Unicode)। কিন্তু প্রতিটি অক্ষরকে আলাদাভাবে সংখ্যায় রূপান্তর করলে শব্দের মূল অর্থ হারিয়ে যায়।

* **অক্ষর-ভিত্তিক সমস্যা:** অক্ষরকে এনকোড করলে কম্পিউটার শুধু বর্ণ চেনে, শব্দের ভাব বা অর্থ বুঝতে পারে না।
* **শব্দ-ভিত্তিক সমাধান (Tokenization):** পুরো শব্দকে একটি 'টোকেন' ধরে সংখ্যায় ম্যাপিং করলে বাক্যের অর্থ ও ব্যাকরণগত সম্পর্ক বজায় থাকে।

In [27]:
# '\u0986' হলো 'আ' অক্ষরের ইউনিকোড মান
c='\u0986'
print(c)

আ


In [28]:
# ord() দিয়ে যেকোনো অক্ষরের ইউনিকোড (সংখ্যা) বের করা হয়
ord('অ')

2437

In [29]:
# chr() দিয়ে ইউনিকোড সংখ্যাকে আবার মূল অক্ষরে রূপান্তর করা হয়
chr(2437)

'অ'

In [30]:
chr(2471)

'ধ'

ধরি আমরা একটা শব্দ বলছি *'কমলচরণ'*। যার অর্থ পদ্মের মতো সুন্দর পা। তিনটাকে আলাদা আলাদা করে ইউনিকোড পয়েন্ট কোড হচ্ছেঃ

In [31]:
import numpy as np
word= 'কমলচরণ'

# প্রতিটি অক্ষরের ord মান নিয়ে array তৈরি
np.array([ord(char) for char in word])

array([2453, 2478, 2482, 2458, 2480, 2467])

সাধারণত পাইথনের বিল্ট-ইন `ord()` ফাংশন একবারে শুধু একটি অক্ষরের মান বের করে (যেমন: `ord('অ')` এর মান 2437)। পুরো NumPy অ্যারের ওপর এটি সরাসরি ব্যবহার করতে গেলে লুপের প্রয়োজন হয়।  পুরো NumPy অ্যারের ওপর একবারে কাজ করানোর জন্যই আমরা চাইলে `np.vectorize(ord)` ব্যবহার করতে পারি, যাতে আলাদা করে কোনো লুপ লিখতে না হয়।  ওয়ান-ডাইমেনশনাল  অ্যারের ক্ষেত্রে সুবিধাটি সামান্য মনে হলেও, ম্যাট্রিক্স বা মাল্টি-ডাইমেনশনাল  ডেটা প্রসেসিংয়ে এটি বেশ কাজে আসবে।

In [32]:
import numpy as np

word = 'কমলচরণ'

# ১. শব্দটির অক্ষরগুলো নিয়ে একটি NumPy অ্যারে তৈরি
# ['ক', 'ম', 'ল', 'চ', 'র', 'ণ']
char_array = np.array(list(word))

# ২. ord ফাংশনটিকে vectorize করা
# নতুন এই v_ord ফাংশনটির ভেতরে আগে থেকেই একটি Internal Loop তৈরি করা থাকে।
v_ord = np.vectorize(ord)

# ৩. এবার সেই ফাংশনটি পুরো অ্যারের ওপর প্রয়োগ করা
ord_array = v_ord(char_array)

print(ord_array)

[2453 2478 2482 2458 2480 2467]


এরকম আরেকটা শব্দ হলো।  চরণকমল: চরণ বা পা-কে পদ্মের সাথে তুলনা করা।

In [33]:
import numpy as np

word = 'চরণকমল'
char_array = np.array(list(word)) # ['চ', 'র', 'ণ','ক', 'ম', 'ল']
v_ord = np.vectorize(ord)

# ফাংশনটি পুরো অ্যারের ওপর প্রয়োগ করা
ord_array = v_ord(char_array)

print(ord_array)


[2458 2480 2467 2453 2478 2482]


NumPy এর মতো আলাদা করে vectorize বা for লুপ লেখার দরকার নেই। TensorFlow-এর unicode_decode স্বয়ংক্রিয়ভাবে পুরো বাংলা স্ট্রিংটিকে একবারে ডিকোড করে ইউনিকোড টেনসর (Tensor)-এ রূপান্তর করে দেয়।

In [34]:
import tensorflow as tf

word = 'কমলচরণ'

# NumPy এর মতো আলাদা vectorize বা for-loop না লিখে সরাসরি
# পুরো টেক্সটকে একবারে ডিকোড করে ইউনিকোড টেনসরে রূপান্তর করার জন্য TensorFlow ব্যবহার করা হয়েছে।
tf_ord_tensor = tf.strings.unicode_decode(word, input_encoding='UTF-8')

print("TF Tensor   :", tf_ord_tensor)

TF Tensor   : tf.Tensor([2453 2478 2482 2458 2480 2467], shape=(6,), dtype=int32)


Output-এ,  

*   `tf.Tensor(...)`: এটি নির্দেশ করে যে আউটপুটটি সাধারণ পাইথন লিস্ট বা NumPy অ্যারে নয়, এটি একটি TensorFlow Tensor।

*   `[2453 2478 2482 2458 2480 2467]`: এটি "কমলচরণ" শব্দটির ৬টি বাংলা অক্ষরের (ক, ম, ল, চ, র, ণ) নিজ নিজ ইউনিকোড (ASCII/Unicode Value) সংখ্যার তালিকা।

*   shape=(6,): এর অর্থ হলো এটি একটি ১D (One-dimensional) অ্যারে বা ভেক্টর, যার মধ্যে মোট ৬টি উপাদান (Element) রয়েছে।

*   dtype=int32: এটি ডেটা টাইপ নির্দেশ করে। প্রতিটি ইউনিকোড সংখ্যা এখানে 32-bit Integer (পূর্ণসংখ্যা) হিসেবে সংরক্ষিত হয়েছে।

# **ইউনিকোড এনকোডিংয়ের লিমিট ও শব্দের এনকোডিংয়ের প্রয়োজনীয়তা:**


একই অক্ষর দিয়ে গঠিত ভিন্ন ভিন্ন শব্দে অক্ষরগুলোর ইউনিকোড মান একই থাকে, কেবল তাদের অবস্থান পরিবর্তিত হয়।

*   কমলচরণঃ `[2453 2478 2482 2458 2480 2467]`
*   চরণকমলঃ `[2458 2480 2467 2453 2478 2482]`



কেবল অক্ষরের এই পজিশনাল বা ইউনিকোড মান দেখে কোনো বাক্যের অর্থ, ভাব বা কনটেক্সট  বোঝা সম্ভব নয়। কারণ ইউনিকোড কেবল অক্ষরের আইডি প্রকাশ করে, ব্যাকরণ বা শব্দের সম্পর্ক নয়।


> *মডেলকে ভাষা বা ভাবের অর্থ বোঝাতে প্রতিটি অক্ষরের বদলে পুরো শব্দকে এনকোড (Word Embedding বা Vectorization) করতে হয়। এর ফলেই মডেল বুঝতে পারে যে 'কমল' এবং 'চরণ' দুটি ভিন্ন শব্দ হওয়া সত্ত্বেও তাদের ভাব বা অর্থ একই।*



তাই সহজ ভাবে বললে বাক্যের ভাবার্থ বের করার জন্য আমাদের অক্ষর এনকোড না করে শব্দকে লেবেল করতে হবে। এই কথা বুঝার জন্য আমরা আগে পুরা একটা বাক্যে কে শব্দ দিয়ে এনকোড করে দেখি।


In [35]:
sentences = 'আমি বই পড়তে পছন্দ করি'

# NumPy এর মতো আলাদা vectorize বা for-loop না লিখে সরাসরি
# পুরো টেক্সটকে একবারে ডিকোড করে ইউনিকোড টেনসরে রূপান্তর করার জন্য TensorFlow ব্যবহার করা হয়েছে।
tf_ord_tensor = tf.strings.unicode_decode(sentences,'UTF-8')

print("TF Tensor   :", tf_ord_tensor)

TF Tensor   : tf.Tensor(
[2438 2478 2495   32 2476 2439   32 2474 2524 2468 2503   32 2474 2459
 2472 2509 2470   32 2453 2480 2495], shape=(21,), dtype=int32)


Output-এ, Shape দেখাচ্ছে 21 মানে বাক্যের ১৭টা অক্ষর ৪ টা স্পেস  হিসেব করে। সাথে নিজ নিজ ইউনিকোড (ASCII/Unicode Value) সংখ্যার তালিকা। আউটপুটে যেখানেই 32 দেখা যাচ্ছে, সেটি মূলত দুটি শব্দের মাঝের স্পেস (' ') নির্দেশ করছে। কারণ স্পেসের Unicode Value হলো ৩২। বাকিটা আগে একবার ব্যাখ্যা করা হয়েছে।

স্পেসের মাধ্যমে শব্দের বিভাজন বোঝা গেলেও, কেবল এই ইউনিকোড সংখ্যাগুলো দেখে বাক্যের কোনো ভাব, ব্যাকরণ বা গভীর অর্থ কিংবা Context Pattern বোঝা সম্ভব নয়।

এখানে আমরা একটা কাজ করতে পারি।  লেবেলিং!


*   'আমি' শব্দকে '1'
*   'বই' শব্দকে '2'
*   'পড়তে' শব্দকে '3'
*   'পছন্দ' হলো '4'
*   'করি' কে '5'

এভাবে লেবেলিং করে দিলাম। আর এই কাজটা করবো tokenizer দিয়ে।

In [36]:
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer(num_words=10, filters='!')

sentences = ['আমি বই পড়তে পছন্দ করি!']



tokenizer.fit_on_texts(sentences)

word_index = tokenizer.word_index
print(word_index)

{'আমি': 1, 'বই': 2, 'পড়তে': 3, 'পছন্দ': 4, 'করি': 5}


# টোকেনাইজার কীভাবে কাজ করলো?
কোড:



```

from tensorflow.keras.preprocessing.text import Tokenizer

```


*   **Keras**: এটি TensorFlow-এর ভেতরের একটি অংশ (High-level API)। এখানে কঠিন কাজগুলো সহজ কোডে লেখার মতো অনেক টুলস আগে থেকেই বানিয়ে রাখা আছে। কমপ্লেক্স লো-লেভেল ম্যাথমেটিক্যাল কোড না লিখে সহজে নিউরাল নেটওয়ার্ক ও ডাটা পাইপলাইন তৈরি করার জন্য এটি ব্যবহার করা হয়।

*  **text প্রসেসিং মডিউল:** Keras- ভিতরে TensorFlow-এর একটি সাব-মডিউল, যা লেখা বা টেক্সট প্রসেস করার জন্য তৈরি কিছু প্রি-বিল্ট ফাংশন ও মেথডের কালেকশন।

*  **Tokenizer ক্লাস:** এটি একটি  পাইথন ক্লাস। এটি মূলত টেক্সট প্রসেসিং ঘরের এমন একটি তৈরি করা টুল, যার প্রধান কাজ হলো যেকোনো লেখা বা বাক্যকে ভেঙে ছোট ছোট টুকরো বা টোকেন কিংবা ইউনিক শব্দে ভাগ করা এবং  এবং প্রতিটি শব্দকে একটি করে নির্দিষ্ট নম্বর বা আইডি দেওয়া হয় যাতে কম্পিউটার বা এআই মডেল সহজে তা বুঝতে পারে।

এক কথায় সহজ অর্থ: `tensorflow.keras.preprocessing.text` থেকে বাক্যকে সংখ্যায় রূপান্তর করার বিশেষ মেশিনটি হল Tokenizer। যা আমরা আমাদের কোডে ব্যবহার করার জন্য import করেছি।


```

tokenizer = Tokenizer(num_words=10, filters='!')

```

**কাজ:** Tokenizer ক্লাসের একটি অবজেক্ট/ইনস্ট্যান্স তৈরি করা হচ্ছে। গুগল কোলাব বা জুপিটার নোটবুকে কোনো ভ্যারিয়েবল যেমন: tokenizer একবার মেমোরিতে তৈরি হয়ে গেলে, আমি স্ক্রিনের যে সেলই রান করি না কেন, `tokenizer.fit_on_texts()` চালু করলেই সে আগের সব হিসেব মেমোরিতে ধরে রাখে। তাই টোকেনাইজারটিকে রি-সেট করার জন্য আবার অবজেক্ট ডিক্লেয়ার করা করেছি।


```

sentences = ['আমি বই পড়তে পছন্দ করি!']

```
একটি পাইথন লিস্ট তৈরি করা হচ্ছে, যেখানে একটি বাংলা বাক্য স্ট্রিং হিসেবে রাখা হয়েছে।

*   **num_words=10:** এটি অবজেক্টের ভেতর সর্বোচ্চ ১০টি Most Frequent শব্দ ধরে রাখার লিমিট সেট করে।

*   **filters='!':** ডিফল্টভাবে টোকেনাইজারে অনেক বিরামচিহ্ন থাকে। এখানে কাস্টম ফিল্টার হিসেবে কেবল '!' দেওয়া হয়েছে। ব্যাকগ্রাউন্ডে টেক্সট প্রসেস হওয়ার সময় প্রতিটি শব্দ থেকে '!' চিহ্নটি রেগুলার এক্সপ্রেশন (Regex) দিয়ে স্ট্রিপ (Strip/Remove) করে ফেলে দেওয়া হয়।





```

tokenizer.fit_on_texts(sentences)

```
1. Lowercase & Clean: বাক্যটিকে ছোট হাতের অক্ষরে রূপান্তর করে এবং filters='!'-এর নিয়ম অনুযায়ী 'করি!' শব্দ থেকে '!' কেটে।  কেবল 'করি' শব্দটা রাখে।

2. Splitting: স্পেস ধরে বাক্যটিকে আলাদা আলাদা শব্দে বা Tokens এ রূপান্তর করে। যেমন: ['আমি', 'বই', 'পড়তে', 'পছন্দ', 'করি']

3. Frequency Counting: প্রতিটি শব্দ পুরো টেক্সটে কতবার এসেছে তার একটি ফ্রিকোয়েন্সি কাউন্টার তৈরি করে।

4. Indexing: সবচেয়ে বেশি ব্যবহৃত শব্দকে 1 থেকে শুরু করে। এরপরের শব্দকে ২। এভাবে পরের শব্দকে 3, তারপরে যেটা বেশি ব্যবহার হয়েছে তাদের সবাইকে এক এক করে ইনডেক্সিং করে নেয়। এখানে সব শব্দ ১ বার করে আসায় বাক্যের ক্রমানুসারে ইনডেক্স তৈরি হবে। যেমনঃ এখানে হবে 'আমি': 1, 'বই': 2, 'পড়তে': 3, 'পছন্দ': 4, 'করি': 5



```

word_index = tokenizer.word_index

```

ভেতরের আসল ঘটনা: `fit_on_texts()` চালানোর সময় টোকেনাইজার নিজের মেমোরিতে প্রতিটি শব্দের জন্য একটা সংখ্যা দিয়ে ডিকশনারি বানিয়ে রাখে।  যেমন: 'আমি': 1, 'বই': 2 ইত্যাদি। এই লাইনটা দিয়ে আমরা মেমোরির সেই প্রস্তুত করা তালিকাটি নিজের একটা ভ্যারিয়েবলে সেভ করে নিচ্ছি, যাতে পরে দেখতে বা ব্যবহার করতে পারি। এটাই word_index।

আউটপুটঃ
```

{'আমি': 1, 'বই': 2, 'পড়তে': 3, 'পছন্দ': 4, 'করি': 5}

```


In [37]:
sentences = [
    'আমি বই পড়তে পছন্দ করি!',
    'আমি বই লিখতে পছন্দ করি'
]


# 'sentenes' এর জায়গায় 'sentences' হবে
tokenizer.fit_on_texts(sentences)

word_index = tokenizer.word_index
print(word_index)

{'আমি': 1, 'বই': 2, 'পছন্দ': 3, 'করি': 4, 'পড়তে': 5, 'লিখতে': 6}



যখন আমরা দুটি বাক্য পাস করলাম
1. 'আমি বই পড়তে পছন্দ করি!'
2. 'আমি বই লিখতে পছন্দ করি'

টোকেনাইজার সব ব্যাকগ্রাউন্ড হিসাব সম্পন্ন করে এই ইনডেক্স তৈরি করলো


```
{'আমি': 1, 'বই': 2, 'পছন্দ': 3, 'করি': 4, 'পড়তে': 5, 'লিখতে': 6}
```
ইনডেক্সিং করার লজিক:

* **ফ্রিকোয়েন্সি (Frequency):**  টোকেনাইজার সবচেয়ে বেশি ব্যবহৃত শব্দগুলোকে আগে সংখ্যা বা আইডি দেয়। এখানে 'আমি', 'বই', 'পছন্দ' ও 'করি', এই ৪টি শব্দ উভয় বাক্যেই ছিল, অর্থাৎ বেশিবার এসেছে। ২ বার করে। তাই টোকেনাইজার এগুলোকে বেশি গুরুত্বপূর্ণ মনে করে সবার আগে ১ থেকে ৪ নম্বর ইনডেক্সিং করেছে।

* **কম ব্যবহৃত শব্দ:** 'পড়তে' এবং 'লিখতে' শব্দ দুটি ১ বার করে এসেছিল। তাই তারা শেষের দিকে ৫ ও ৬ নম্বর ইনডেক্স পেয়েছে।

In [41]:
tokenizer.word_counts

OrderedDict([('আমি', 3),
             ('বই', 3),
             ('পড়তে', 2),
             ('পছন্দ', 3),
             ('করি', 3),
             ('লিখতে', 1)])

টেক্সটে শব্দগুলো যেভাবে একের পর এক প্রথম ইনপুট দেওয়া হয়েছে, এখানেও সেগুলোকে ঠিক সেই একই অনুক্রমে সংরক্ষণ করা এই হলো পাইথনের এই ডিকশনারির কাজে। যা OrderedDic নামে দেখাচ্ছে।

```
OrderedDict([('আমি', 3),
             ('বই', 3),
             ('পড়তে', 2),
             ('পছন্দ', 3),
             ('করি', 3),
             ('লিখতে', 1)])
```

এর মানে হলো প্রতিটি শব্দের পাশে সেই শব্দটি কতবার এসেছে, সেটা রাখা হয়েছে। যেমন:


  আমি → 3 বার  
  বই → 3 বার  
পড়তে → 2 বার  
পছন্দ → 3 বার  
করি → 3 বার  
লিখতে → 1 বার

'আমি বই পড়তে পছন্দ করি!'  
'আমি বই লিখতে পছন্দ করি'


এখানে প্রথম বাক্যের শব্দ গুলো দ্বিতীয় বাক্যে প্রায় সব আছে শুধু 'লিখতে' শব্দটা ছাড়া। উপরে আরেকটা সেলো প্রথম বাক্যটা ছিলো। তাই 3 করে দেখাচ্ছে। তার মানে শেষ শব্দ এখানে 'লিখতে' তাও এটার স্থান অর্ডার ডিকশনারির শেষে অবস্থান।  অর্থাৎ লেখার মধ্যে কোন শব্দ কতবার এসেছে, তার word frequency/count বের করে OrderedDict এ রাখা হয়েছে। শব্দ যুক্ত হওয়ার ক্রম অনুসারে।

In [42]:
tokenizer.word_index

{'আমি': 1, 'বই': 2, 'পছন্দ': 3, 'করি': 4, 'পড়তে': 5, 'লিখতে': 6}

word_index নিয়ে শুরুতে বলেছিলাম। আবার বলছি fit_on_texts() চালানোর সময় টোকেনাইজার নিজের মেমোরিতে প্রতিটি শব্দের জন্য একটা সংখ্যা দিয়ে ডিকশনারি বানিয়ে রাখে।  যেমন: 'আমি': 1, 'বই': 2 ইত্যাদি। এই লাইনটা দিয়ে আমরা মেমোরির সেই প্রস্তুত করা তালিকাটি নিজের একটা ভ্যারিয়েবলে সেভ করে নিয়েছিলাম, যাতে পরে দেখতে বা ব্যবহার করতে পারি। এটাই word_index।  উপরে টোকেন অবজেক্টের word_index টা দেখানো হয়েছে।

In [39]:
tokenizer.get_config()

{'num_words': 10,
 'filters': '!',
 'lower': True,
 'split': ' ',
 'char_level': False,
 'oov_token': None,
 'document_count': 3,
 'word_counts': '{"\\u0986\\u09ae\\u09bf": 3, "\\u09ac\\u0987": 3, "\\u09aa\\u09dc\\u09a4\\u09c7": 2, "\\u09aa\\u099b\\u09a8\\u09cd\\u09a6": 3, "\\u0995\\u09b0\\u09bf": 3, "\\u09b2\\u09bf\\u0996\\u09a4\\u09c7": 1}',
 'word_docs': '{"\\u09ac\\u0987": 3, "\\u0986\\u09ae\\u09bf": 3, "\\u09aa\\u09dc\\u09a4\\u09c7": 2, "\\u0995\\u09b0\\u09bf": 3, "\\u09aa\\u099b\\u09a8\\u09cd\\u09a6": 3, "\\u09b2\\u09bf\\u0996\\u09a4\\u09c7": 1}',
 'index_docs': '{"2": 3, "1": 3, "3": 3, "5": 2, "4": 3, "6": 1}',
 'index_word': '{"1": "\\u0986\\u09ae\\u09bf", "2": "\\u09ac\\u0987", "3": "\\u09aa\\u099b\\u09a8\\u09cd\\u09a6", "4": "\\u0995\\u09b0\\u09bf", "5": "\\u09aa\\u09dc\\u09a4\\u09c7", "6": "\\u09b2\\u09bf\\u0996\\u09a4\\u09c7"}',
 'word_index': '{"\\u0986\\u09ae\\u09bf": 1, "\\u09ac\\u0987": 2, "\\u09aa\\u099b\\u09a8\\u09cd\\u09a6": 3, "\\u0995\\u09b0\\u09bf": 4, "\\u09aa\\

# `get_config()` মেথড
এটি মূলত টোকেনাইজারের ভেতরের সমস্ত গোপনীয় তথ্য বা স্টেট এক নজরে বের করে আনে:  

**document_count** → মোট ২টি আলাদা বাক্য প্রসেস করা হয়েছে।  
'আমি বই পড়তে পছন্দ করি!'  
'আমি বই লিখতে পছন্দ করি'

**word_counts** → পুরো টেক্সটে কোন শব্দ কতবার এসেছে তার সংখ্যা। ইউনিকোড হিসেবে ডিকোড করা, যেমন: \\u0986\\u09ae\\u09bf = 'আমি', যার কাউন্ট ৩।

**word_docs** → একটি শব্দ মোট কতটি ডুকুমেন্ট এ এসেছে তা রিটার্ন করে।  যেমন: 'বই' ৩টি বাক্যে এসেছে।

**index_word** → ইনডেক্স থেকে শব্দ খোঁজার ম্যাপ ("1": "আমি").

**word_index** → শব্দ থেকে ইনডেক্স খোঁজার ম্যাপ ("পছন্দ":3).

**filters** '!' $\rightarrow$ কেবল '!' চিহ্নটিকে ফিল্টার বা বাদ দেওয়ার নিয়ম হিসেবে সেট করা হয়েছে।

**char_level:** False $\rightarrow$ ক্যারেক্টার ধরে নয়, এটি শব্দ ধরে টোকেনাইজ করছে।

**split:** ' ' $\rightarrow$ বাক্যকে শব্দে ভাগ করার জন্য 'স্পেস' (Space) ক্যারেক্টারকে সেপারেটর হিসেবে ব্যবহার করছে।

---
### 💡 Extra: পাইথনে ইউনিকোড ডিকোড করার নিয়ম

> **নোট:**  
> Keras Tokenizer অনেক সময় বাংলা বা অন্যান্য অ-ইংরেজি টেক্সটকে প্রসেস করার পর এমন ইউনিকোড ফরম্যাটে (`\u0986\u09ae...`) আউটপুট দেয়। এটি কোর পাইথনের `json` মডিউল ব্যবহার করে সহজে পড়ার যোগ্য বাংলা টেক্সটে ডিকোড করে নেওয়া যায়।

In [47]:
# Extra: পাইথনে ইউনিকোড কে ডিকোড করে দেখতে চাইলে:
import json

data = '{"\\u0986\\u09ae\\u09bf": 3}'
print(json.loads(data))
# আউটপুট দেখাবে: {'আমি': 3}

{'আমি': 3}


পাইথনে নিজে ডিকোড করার সহজ কোড এটি। নিচে পুরা সব word_counts দিয়ে করা দেখানো হলো

In [48]:
import json

data = {
    'word_counts': '{"\\u0986\\u09ae\\u09bf": 3, "\\u09ac\\u0987": 3, "\\u09aa\\u09dc\\u09a4\\u09c7": 2, "\\u09aa\\u099b\\u09a8\\u09cd\\u09a6": 3, "\\u0995\\u09b0\\u09bf": 3, "\\u09b2\\u09bf\\u0996\\u09a4\\u09c7": 1}',
    'index_word': '{"1": "\\u0986\\u09ae\\u09bf", "2": "\\u09ac\\u0987", "3": "\\u09aa\\u099b\\u09a8\\u09cd\\u09a6", "4": "\\u0995\\u09b0\\u09bf", "5": "\\u09aa\\u09dc\\u09a4\\u09c7", "6": "\\u09b2\\u09bf\\u0996\\u09a4\\u09c7"}'
}

# json.loads() ব্যবহার করলেই এটি ডিকোড হয়ে যাবে
decoded_word_counts = json.loads(data['word_counts'])
print(decoded_word_counts)

{'আমি': 3, 'বই': 3, 'পড়তে': 2, 'পছন্দ': 3, 'করি': 3, 'লিখতে': 1}


---
## 🔍 Deep Dive: ব্যাকগ্রাউন্ডে Tokenizer কীভাবে কাজ করে?

> 💡 **আমার মতো কৌতূহলী মনের জন্য (Behind the Scenes):**  
> কোনো থার্ড-পার্টি লাইব্রেরি বা এক্সট্রা মডিউল (যেমন: Keras/NLTK) ছাড়াই সম্পূর্ণ **কোর পাইথন (Core Python)** ব্যবহার করে `fit_on_texts()` মেথডটির ভেতরের মেকানিজম নিচে সহজ কোডে দেখানো হলো।

In [51]:
# ১. কাঁচা টেক্সট (Raw Sentences)
sentences = ['আমি বই পড়তে পছন্দ করি', 'আমি বই লিখতে পছন্দ করি']

# ২. Tokenizer-এর ভেতরের লজিক (সিম্পল পাইথন কোড)
word_counts = {}

for sentence in sentences:
    words = sentence.split()  # স্পেস দিয়ে শব্দে ভাগ করা
    for word in words:
        # ব্যাকগ্রাউন্ডে শব্দের সংখ্যা গণনা করা
        word_counts[word] = word_counts.get(word, 0) + 1

# ৩. ফ্রিকোয়েন্সি অনুযায়ী ইনডেক্সিং করা
# বেশি ব্যবহৃত শব্দগুলোকে ১, ২, ৩... ইনডেক্স দেওয়া
sorted_words = sorted(word_counts, key=word_counts.get, reverse=True)
word_index = {word: index + 1 for index, word in enumerate(sorted_words)}

print("word_counts:", word_counts)
print("word_index:", word_index)

word_counts: {'আমি': 2, 'বই': 2, 'পড়তে': 1, 'পছন্দ': 2, 'করি': 2, 'লিখতে': 1}
word_index: {'আমি': 1, 'বই': 2, 'পছন্দ': 3, 'করি': 4, 'পড়তে': 5, 'লিখতে': 6}


# ১.  Sentences

`sentences = ['আমি বই পড়তে পছন্দ করি', 'আমি বই লিখতে পছন্দ করি']`

* ইনপুট হিসেবে প্রসেস করার জন্য মূল বাক্য দুটিকে একটি লিস্টের (List) ভেতরে নেওয়া হয়েছে।

---

# ২. Tokenizer-এর ভেতরের লজিক (সিম্পল পাইথন কোড)

`word_counts = {}`

* শব্দগুলোর সংখ্যা বা ফ্রিকোয়েন্সি জমা রাখার জন্য একটি খালি ডিকশনারি তৈরি করা হয়েছে।

`for sentence in sentences:`

* লিস্টের ভেতর থেকে প্রতিটি বাক্যকে এক এক করে লুপের মাধ্যমে প্রসেস করা হচ্ছে।

`words = sentence.split()`

* `split()` মেথড ব্যবহার করে বাক্যের স্পেস ধরে ধরে শব্দগুলোকে আলাদা করা হচ্ছে।

`for word in words:`

* আলাদা করা শব্দগুলোর ওপর আরও একটি লুপ চালিয়ে প্রতিটি শব্দকে আলাদাভাবে ট্র্যাক করা হচ্ছে।

`word_counts[word] = word_counts.get(word, 0) + 1`

* `get()` মেথড দিয়ে শব্দটি ডিকশনারিতে আগে থেকে আছে কি না তা চেক করা হচ্ছে। যদি থাকে তবে তার আগের মানের সাথে ১ যোগ হচ্ছে, আর না থাকলে মান হিসেবে ১ বসানো হচ্ছে।

---

# ৩. ফ্রিকোয়েন্সি অনুযায়ী ইনডেক্সিং করা

`sorted_words = sorted(word_counts, key=word_counts.get, reverse=True)`

* সবচেয়ে বেশি ব্যবহৃত শব্দগুলোকে ক্রমানুসারে (Descending Order) সাজিয়ে নেওয়া হচ্ছে।

`word_index = {word: index + 1 for index, word in enumerate(sorted_words)}`

* `enumerate()` ব্যবহার করে প্রতিটি শব্দকে ১ থেকে শুরু করে একটি  ইউনিক আইডি নম্বর দেওয়া হচ্ছে।

---

# ৪. print

`print("word_counts:", word_counts)`  
`print("word_index:", word_index)`

* ব্যাকগ্রাউন্ডে গণনা করা শব্দের পরিমাণ এবং তৈরি হওয়া ইনডেক্স ম্যাপটি টার্মিনালে আউটপুট হিসেবে দেখাচ্ছে।